clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaDirRequest.java

```diff
@@ -0,0 +1,148 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.requests;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.utils.CollectionUtils;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class AlterReplicaDirRequest extends AbstractRequest {
+
+    // request level key names
+    private static final String LOG_DIRS_KEY_NAME = "log_dirs";
+
+    // log dir level key names
+    private static final String LOG_DIR_KEY_NAME = "log_dir";
+    private static final String TOPICS_KEY_NAME = "topics";
+
+    // topic level key names
+    private static final String TOPIC_KEY_NAME = "topic";
+    private static final String PARTITIONS_KEY_NAME = "partitions";
+
+    private final Map<TopicPartition, String> partitionDirs;
+
+    public static class Builder extends AbstractRequest.Builder<AlterReplicaDirRequest> {
+        private final Map<TopicPartition, String> partitionDirs;
+
+        public Builder(Map<TopicPartition, String> partitionDirs) {
+            super(ApiKeys.ALTER_REPLICA_DIR);
+            this.partitionDirs = partitionDirs;
+        }
+
+        @Override
+        public AlterReplicaDirRequest build(short version) {
+            return new AlterReplicaDirRequest(partitionDirs, version);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder builder = new StringBuilder();
+            builder.append("(type=AlterReplicaDirRequest")
+                .append(", partitionDirs=")
+                .append(partitionDirs)
+                .append(")");
+            return builder.toString();
+        }
+    }
+
+    public AlterReplicaDirRequest(Struct struct, short version) {
+        super(version);
+        partitionDirs = new HashMap<>();
+        for (Object logDirStructObj : struct.getArray(LOG_DIRS_KEY_NAME)) {
+            Struct logDirStruct = (Struct) logDirStructObj;
+            String logDir = logDirStruct.getString(LOG_DIR_KEY_NAME);
+            for (Object topicStructObj : logDirStruct.getArray(TOPICS_KEY_NAME)) {
+                Struct topicStruct = (Struct) topicStructObj;
+                String topic = topicStruct.getString(TOPIC_KEY_NAME);
+                for (Object partitionObj : topicStruct.getArray(PARTITIONS_KEY_NAME)) {
+                    int partition = (Integer) partitionObj;
+                    partitionDirs.put(new TopicPartition(topic, partition), logDir);
```

Comment:
I think we don't need to add this logic in `AlterReplicaDirRequest` because the current implementation will always provide unique partition and log when instantiating `AlterReplicaDirRequest`. `AlterReplicaDirRequest` will only be instantiated in `AdminClient.alterReplicaDir()`. Because `AdminClient.alterReplicaDir()` takes `Map<TopicPartitionReplica, String> replicaAssignment` as input, it guarantees that the partition in the `AlterReplicaDirRequest` will be specified only once.

Also, `AlterReplicaDirRequest.toStruct()` will group `TopicPartitionReplica` by logDir and thus the logDir in the `AlterReplicaDirRequest` will be specified only once.

Reply:
Yes, the java client behaves as you described. I was mostly concerned about how the broker handles requests from non-java clients.

Reply:
Thanks for the comment @junrao !

By non-Java clients, do you mean the clients written by third-party and not maintained by in Apache Kafka repository? Here is my thought:

- If a thirty party client constructs the `AlterReplicaDirRequest` using the API `AlterReplicaDirRequest.Builder(Map<TopicPartition, String> partitionDirs)`, we still guarantee that both log directory and the topicPartition in the resulting `AlterReplicaDirRequest` will be unique. Thus server doesn't have to worry about it.

- If a third party client constructs `AlterReplicaDirRequest` without using our builder, and it constructs `AlterReplicaDirRequest` with duplicated topicPartition by mistake, the server is protected won't be affected by this. This is because the `AlterReplicaDirRequest(Struct struct, short version)` will generate the `Map<TopicPartition, String> partitionDirs` which contains well-defined partition to logDir mapping. Also, since user constructs `AlterReplicaDirRequest` without using our builder, we won't help user detect this mistake by changing the code here.

BTW, this issue seems to also exist in `AlterConfigsRequest`, which may contain duplicated entries for the same resource type and resource name. It may be reasonable to handle them in the same way as this patch does.

Does it make sense?

Reply:
Ok, if doesn't do harm on the server, we can punt on this.

Reply:
@lindong28, Jun was referring to clients that don't use our code at all. librdkafka, kafka-topics, etc.

Reply:
@ijuma I see. Thanks for the information.